[RPC Framework] Support remote device format "<workername>/<device>"

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46773 [RPC Framework] Support remote device format "<workername>/<device>"**

Changed the constructor of RemoteModule to accept a `remote_device` arg in the following format:
"<workername>/<device>" (e.g., "trainer0/cpu", "ps0/cuda:0")

This arg merges the original `on` and `device` arg.

Original PR issue: RemoteDevice Format #46554

Differential Revision: [D24482562](https://our.internmc.facebook.com/intern/diff/D24482562/)

Commit:
Support remote device format "<workername>/<device>".

Changed the constructor of RemoteModule to accept a `remote_device` arg in the following format:
"<workername>/<device>" (e.g., "trainer0/cpu", "ps0/cuda:0")

This arg merges the original `on` and `device` arg.

Differential Revision: [D24482562](https://our.internmc.facebook.com/intern/diff/D24482562/)

[ghstack-poisoned]

pritamdamania87 commented:
Would be nice to mention the original issue in the PR summary: https://github.com/pytorch/pytorch/issues/46554

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -155,18 +155,22 @@ def __init__(
         kwargs = kwargs if kwargs is not None else {}
 
-        self.on = on
+        [self.on, self.device] = remote_device.split("/")
```
If this split is invalid, the user would get an error like "too many values to unpack". Ideally, we want to raise an error like this to the user if the format is not valid: "Couldn't parse remote_device: {}, the accepted format is <workername>/<device>".format(remote_device)". This makes it much easier for the user to understand what went wrong.

In addition to this, specifying just "workername" should be allowed as well. In this case, we assume the default is "cpu".

pritamdamania87 commented on torch/testing/_internal/distributed/ddp_under_dist_autograd_test.py:
```diff
@@ -667,7 +667,7 @@ def test_ddp_dist_autograd_local_vs_remote_gpu(self):
 
         remote_layer1 = RemoteModule(
-            "worker0", device="cpu", module_cls=nn.Linear, args=(10, 7, False)
+            remote_device="worker0/cpu", module_cls=nn.Linear, args=(10, 7, False)
```
Let's add a test for just "worker0" as well.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -108,20 +108,21 @@ def test_bad_module(self):
             r"Expect `module_cls\(\*args, \*\*kwargs\)` returns an instance of <class nn.Module>,",
         ):
-            RemoteModule(dst_worker_name, "cpu", BadModule, args, kwargs)
+            RemoteModule(remote_device, BadModule, args, kwargs)
```
We should add tests for invalid remote devices, ex:

1. "trainer0/"
2. "trainer0/xyz"
3. "trainer0/cuda:0/cuda:1"
4. "/"
5. "/cuda:0"

SciPioneer replied:
Thanks for the suggestion! Incorporated these test cases with `test_invalid_devices`, which has already covered case 2 you listed here.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -155,18 +155,22 @@ def __init__(
         kwargs = kwargs if kwargs is not None else {}
 
-        self.on = on
+        [self.on, self.device] = remote_device.split("/")
```
Let's add a helper method `_parse_remote_device` under `torch.distributed.rpc.utils` which we can later reuse in other parts like RemoteDevice etc.

SciPioneer replied:
Created a utils.py under torch/distributed/rpc.

pritamdamania87 reviewed (commented):
Thanks for getting this PR up quickly!

Commit:
Update on "[RPC Framework] Support remote device format "<workername>/<device>""


Changed the constructor of RemoteModule to accept a `remote_device` arg in the following format:
"<workername>/<device>" (e.g., "trainer0/cpu", "ps0/cuda:0")

This arg merges the original `on` and `device` arg.

Differential Revision: [D24482562](https://our.internmc.facebook.com/intern/diff/D24482562/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -102,6 +103,7 @@ def __init__(
             remote_device (str): Device on the destination worker where we‘d like to place this module.
                 The format should be "<workername>/<device>", where the device field can be parsed as torch.device type.
                 E.g., "trainer0/cpu", "ps0/cuda:0".
```
nit: add "trainer0" to the example as well.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -102,6 +103,7 @@ def __init__(
             remote_device (str): Device on the destination worker where we‘d like to place this module.
                 The format should be "<workername>/<device>", where the device field can be parsed as torch.device type.
                 E.g., "trainer0/cpu", "ps0/cuda:0".
+                In addition, the device field can be optional, and the default value is "cpu".
```
nit: no comma before and.

Commit:
Update on "[RPC Framework] Support remote device format "<workername>/<device>""


Changed the constructor of RemoteModule to accept a `remote_device` arg in the following format:
"<workername>/<device>" (e.g., "trainer0/cpu", "ps0/cuda:0")

This arg merges the original `on` and `device` arg.

Differential Revision: [D24482562](https://our.internmc.facebook.com/intern/diff/D24482562/)

[ghstack-poisoned]

SciPioneer commented:
> Would be nice to mention the original issue in the PR summary: #46554

Thanks for the reminder!
